daemon/cluster/cluster.go

```diff
@@ -871,6 +872,34 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
 	return services, nil
 }
 
+// imageWithDigestString takes an image such as name or name:tag
+// and returns the image pinned to a digest, such as name@sha256:34234...
+func (c *Cluster) imageWithDigestString(ctx context.Context, image, encodedAuth string) (string, error) {
+	// TODO(nishanttotla): create AuthConfig from encodedAuth
+
+	ref, err := reference.ParseNamed(image)
+	if err != nil {
+		logrus.Warnf("unable to parse image name: %s", err.Error())
+		return "", err
+	}
+	// only pin if not a canonical reference (i.e. with digest)
+	if _, ok := ref.(reference.Canonical); !ok {
+		// add default tag if not present
+		if reference.IsNameOnly(ref) {
+			ref = reference.WithDefaultTag(ref)
+		}
+		namedTaggedRef, _ := ref.(reference.NamedTagged)
+		dgst, err := c.config.Backend.ResolveTagToDigest(ctx, namedTaggedRef, &apitypes.AuthConfig{})
+		if err != nil {
+			logrus.Warnf("unable to retrieve digest for image: %s", err.Error())
+			return "", err
+		}
+		logrus.Infof("pinned image to digest: %s", dgst)
```

Comment:
This should probably be a `Debugf`, and ideally should provide more context about what was pinned. I'd suggest creating a logger at the top of this function with `logrus.WithFields`. Then if you use that logger for the log messages, they can have the service name, image reference, and so on filled in.